### PR TITLE
Allow plural outputs to be cached

### PR DESCRIPTION
**Pull request by lptr**
When `@OutputFiles` or `@OutputDirectories` is used with an `Iterable` (instead of a `Map`), we previously disabled caching without much explanation. The original reason for this was that we expected a difference in cardinality of outputs between the task producing the output and the one consuming it. However, if we represent each entry in the collection as an indexed property, then unmatched cardinality will produce a different cache key anyway. Thus there's no need for this limitation.

### CODE REVIEW

**Comment by lptr** [comment]
Branch build: https://builds.gradle.org/viewQueued.html?itemId=15290212

File: subprojects/core/src/integTest/groovy/org/gradle/api/internal/tasks/TaskCacheabilityReasonIntegrationTest.groovy [hunk 1]
```diff
@@ -133,20 +133,22 @@ class TaskCacheabilityReasonIntegrationTest extends AbstractIntegrationSpec impl
...
+                }
                 
                 @TaskAction
                 void generate() {}
             }
             
             task cacheable(type: PluralOutputs) {
                 cachingEnabled = false
-                disabledReason = "Declares multiple output files for the single output property 'outputFiles' via `@OutputFiles`, `@OutputDirectories` or `TaskOutputs.files()`"
+                disabledReason = "Declares complex output for the output property 'outputFiles'"
```
Comment (wolfs) [inline subprojects/core/src/integTest/groovy/org/gradle/api/internal/tasks/TaskCacheabilityReasonIntegrationTest.groovy]: That will probably require an update in build scans as well. Currently, the reason is "Multiple outputs declared": 
image

I didn't find any help text for it, I guess we would need to add a section to the user guide or the build cache guide explaining what kind of outputs we support for caching and link to it. WDYT?

Comment (wolfs) [reply subprojects/core/src/integTest/groovy/org/gradle/api/internal/tasks/TaskCacheabilityReasonIntegrationTest.groovy]: I also think that `complex` is very vague. Can we be more specific in what is allowed? Something like "Only files, directories and unfiltered file trees are allowed"?

File: subprojects/docs/src/docs/release/notes.md [hunk 2]
```diff
@@ -58,6 +58,14 @@ In this way, stale execution data, which would cause non-repeatable task outputs
 Since Gradle now takes care of removing the execution data, the `JacocoPluginExtension.append` property has been deprecated.
 The JaCoCo agent is always configured with `append = true`, so it can be used when running tests in parallel. 
 
+### Plural task output properties don't disable caching anymore
```
Comment (wolfs) [inline subprojects/docs/src/docs/release/notes.md]: I think we also need to add something/change the user guide and adjust the Javadocs on `OutputFiles` and `OutputDirectories`.

**Comment by lptr** [comment]
I've added a new non-caching reason `NON_CACHEABLE_TREE_OUTPUT`. 

I also disabled caching if any tree appears in the output for the following reasons:

* it's easier to explain to users,
* it's unlikely that people would want to use an unfiltered `DirectoryFileTree` (which is an internal type anyway) as an output,
* we don't support `DirectoryFileTree` as an `@OutputDirectory` either (we actually throw an exception).

I've updated our horrible user guide chapter about task inputs with the changes.

**Comment by lptr** [comment]
Branch build: https://builds.gradle.org/viewQueued.html?itemId=15343943

**Comment by lptr** [comment]
This will need to be mentioned in https://github.com/gradle/dotcom/issues/1399, too.

**Comment by lptr** [comment]
Branch build: https://builds.gradle.org/viewQueued.html?itemId=15349354

**Comment by lptr** [comment]
Based on @facewindu's feedback I removed the old enum value.

**Pull request merged**